drivers/timer/xtensa_sys_timer.c

```diff
@@ -99,14 +101,175 @@ static u32_t __noinit cycles_per_tick;
 static u32_t __noinit max_system_ticks;
 static u32_t idle_original_ticks;
 static u32_t __noinit max_load_value;
+
+#ifdef CONFIG_TICKLESS_KERNEL
+static u32_t last_timer_value;
+#else
 static unsigned char timer_mode = TIMER_MODE_PERIODIC;
 static unsigned char idle_mode = IDLE_NOT_TICKLESS;
+#endif	/* CONFIG_TICKLESS_KERNEL */
+
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/**
+ * @brief provides total programmed.
+ *
+ * returns : total number of sys ticks programmed.
+ */
+
+uint32_t _get_program_time(void)
+{
+	return idle_original_ticks;
+}
+
+/**
+ * @brief : Provides Remaining Time for timer to expire.
+ *
+ * returns : time in Ticks to expire.
+ */
+
+uint32_t _get_remaining_program_time(void)
+{
+	u32_t c; /* Current time (time within this function execution) */
+	u32_t f; /* Idle timer programmed fire time */
+	u32_t r; /*reamining time to the timer to expire */
+
+	if (!idle_original_ticks) {
+		return 0;
+	}
+
+	f = GET_TIMER_FIRE_TIME();	/* Time to Fire */
+	c = GET_TIMER_CURRENT_TIME(); /* Current Time */
+	r = f > c ? (f - c) / cycles_per_tick : 0;
+	return r;
+}
+
+/**
+ * @brief provides total ticks elapsed.
+ *
+ * returns : total number of timer ticks passed since last  programming.
+ */
+uint32_t _get_elapsed_program_time(void)
+{
+	if (!idle_original_ticks) {
+		return 0;
+	}
+	return (_get_program_time() - _get_remaining_program_time());
+}
+
+/**
+ * @brief provides time remaining to reach timer overflow.
+ *
+ * This function returns how many timer Cycles remaining for timer to overflow.
+ * This will be required when we will program timer compare value to maximum
+ * possible value.
+ *
+ * returns : difference between current timer cycle and Maximum possible cycle.
+ */
+int32_t _get_max_clock_time(void)
+{
+	u32_t C;
+
+	C = GET_TIMER_CURRENT_TIME(); /* Current Timer Cycle */
+	return (MAX_TIMER_CYCLES - C);
+}
+
+/* @brief Programe Timer to Maximum Value */
+static inline void _set_max_clock_time(void)
+{
+	int key;
+
+	key = irq_lock();
```

Comment:
Any particular reason why we are making this atomic? I don't see irq_lock and unlock being used in the  original xtensa_sys_timer.c file where GET_TIMER_CURRENT_TIME is being referred.

Reply:
good point @rgundi, I was also thinking of avoiding it. But I have used irq_lock at _sys_clock_tick_count" (system level time keeping) updates. 
I wanted to keep writing operation inside _sys_clock_tick_count to be atomic so that we do not get erroneous value.